### DISCUSSION TITLE
Question: How to add WorkflowId to each log

### DISCUSSION THREAD

**Discussion by alkampfergit** [elsa-workflows/elsa-core#4027] [Q&A]
I was not able to find an answer to this question. I'm using serilog and with Enrichers we can simply add extra info on logging. It would be nice to add workflowId for each log that happens in a context of a workflow execution, But I was not able to find anything in elsa that gives me a WorkflowContext if I'm in the context of a running worfklow or null if I'm outside the context. 

Where should I look? is AmbientActivityExecutionContext the correct class to use?

Thanks.

**Comment by sfmskywalker** [comment]
This is for Elsa 3:

A good place to add logging context would be in the form of a workflow execution middleware component - these work conceptually the same as ASP.NET middleware.

From such a middleware, you have access to the workflow execution context, from which you can add the workflow instance ID and workflow definition ID, for example, to some context.

I'm a little bit rusty on SeriLog, but your middleware component could look like this:

```csharp
public class WorkflowContextLoggingMiddleware : WorkflowExecutionMiddleware
{
   public override async ValueTask InvokeAsync(WorkflowExecutionContext context)
   {
      using (LogContext.PushProperty("WorkflowId", context.Id))
      {
         // Invoke next middleware.
         await Next(context);
      }
   }
}
```

Elsa 2 doesn't have middleware, so we would have to do something similar to what we did with `AmbientActivityExecutionContext` that you mentioned. Although you could use `AmbientActivityExecutionContext`, it will be change for each activity being executed. But maybe this is fine. In any case, it does give you access to the workflow execution context.

**Comment by alkampfergit** [comment]
Thank you for providing the information. I understand that we will be transitioning to Elsa3 in the future. When that happens, I will make sure to update my code accordingly. Currently, my code works well with the AmbientActivityExecutionContext encapsulated within a Serilog Enricher. However, I must say, the approach being used in Elsa3 appears to be significantly improved and more efficient.
